Sources/SwiftFormatWhitespaceLinter/WhitespaceLinter.swift

```diff
@@ -0,0 +1,332 @@
+// swift-tools-version:4.2
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Formatter open source project.
+//
+// Copyright (c) 2018 Apple Inc. and the Swift Formatter project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Swift Formatter project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftFormatCore
+import SwiftSyntax
+
+/// Emits linter errors for whitespace style violations by comparing the raw text of the input Swift
+/// code with formatted text.
+public struct WhitespaceLinter {
+
+  /// The text of the input source code to be linted.
+  let userText: String
+
+  /// The formatted version of `userText`.
+  let formattedText: String
+
+  /// The Context object containing the DiagnosticEngine.
+  let context: Context
+
+  /// Creates a new WhitespaceLinter with the given context.
+  ///
+  /// - Parameters:
+  ///   - user: The text of the Swift source code to be linted.
+  ///   - formatted: The formatted text to compare to `user`.
+  ///   - context: The context object containing the DiagnosticEngine instance we wish to use.
+  public init(user: String, formatted: String, context: Context) {
+    self.userText = user
+    self.formattedText = formatted
+    self.context = context
+  }
+
+  /// Perform whitespace linting.
+  public func lint() {
+    var userOffset = 0
+    var formOffset = 0
+    var isFirstCharater = true
+    var lastChar: Character?
+
+    repeat {
+      let userNext = nextCharacter(offset: userOffset, data: self.userText)
+      let formNext = nextCharacter(offset: formOffset, data: self.formattedText)
+
+      // `userText` and `formattedText` should only differ in their whitespace characters.
+      if userNext.char != formNext.char {
+        fatalError("Characters do not match")
+      }
+
+      lastChar = userNext.char
+
+      compareWhitespace(
+        userOffset: userOffset,
+        isFirstCharacter: isFirstCharater,
+        userWs: userNext.whitespace,
+        formattedWs: formNext.whitespace
+      )
+
+      userOffset = userNext.offset + 1
+      formOffset = formNext.offset + 1
+      isFirstCharater = false
+    } while lastChar != nil
+  }
+
+  /// Compare the whitespace buffers between the user text and formatted text, and emit linter
+  /// errors accordingly.
+  ///
+  /// Note: properly formatted whitespace will always be some number of newlines characters,
+  /// followed by some number of spaces. e.g. "\n ", "\n\n  ", "\n", " ". The user's whitespace
```

Comment:
For our future selves (or future others), it might be helpful to add "because the pretty printer ensures there is no trailing whitespace".

Reply:
Done.